nixos/step-ca: Declarative step-ca service

mohe2015 opened the pull request:
Previous conversation participants: @tashian @midchildan @aanderse

###### Motivation for this change
Create a **declarative** step-ca module instead of a more or less imperative one (which is likely hard to use and maintain using nix).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

mohe2015 committed:
maintainers: add mohe2015

mohe2015 committed:
nixos/step-ca: Add declarative step-ca service

Co-authored-by: midchildan <git@midchildan.org>

r-rmcgibbo commented:
Result of `nixpkgs-review pr 112322` at 91acf595 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li><a href="https://gist.github.com/665f67d258d766fbb59f8dba67602052">step-ca</a></li></li>
  </ul>
</details>

r-rmcgibbo commented:
Result of `nixpkgs-review pr 112322` at 91acf595 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li><a href="https://gist.github.com/3236113ff489c06e60124486cfcb9da6">step-ca</a></li></li>
  </ul>
</details>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/typechecking-options-based-on-a-key/11517/1

midchildan commented:
Thanks for your work! I think this PR is in better shape than the previous one as it would support a wider range of use cases.

I believe we can do with fewer suboptions though, considering that the easiest way to configure this module would be to generate it using `step ca init` and import it using `builtins.fromJSON`. So instead of trying to describe every option in `ca.json`, it'd be more helpful if we document how to configure this module in `settings.description`. Reducing the number of suboptions would also improve maintainability, because keeping the options accurate and in sync with upstream would require constant effort.

> How can I type the provisioners option where the type option defines the available other options. 

You could create custom types, but I think `listOf anything` is enough.

Church- commented:
Yeah this does appear better and easier to use then the previous module we worked on @mohe2015

Thumbsup from me.

midchildan commented on nixos/modules/services/security/step-ca.nix:
```diff
@@ -0,0 +1,68 @@
+      enable = lib.mkEnableOption "the smallstep certificate authority server";
+      settings = lib.mkOption {
+        type = lib.types.anything;
+        description = "See https://smallstep.com/docs/step-ca/configuration. https://smallstep.com/docs/step-cli/basic-crypto-operations#run-an-offline-x509-certificate-authority may also be useful. The easiest way is probably to use the existing smallstep tools to set up the CA and then import the config using e.g. `builtins.fromJSON`";
```
The [docbook](https://docbook.org) markup should be used for documenting NixOS modules.

```suggestion
        description = ''
          Settings that goes into <filename>ca.json</filename>. See
          <link xlink:href="https://smallstep.com/docs/step-ca/configuration">
          the step-ca manual</link> for more information. The easiest way to
          configure this module would be to run <literal>step ca init</literal>
          to generate <filename>ca.json</filename> and then import it using
          <literal>builtins.fromJSON</literal>.
          <link xlink:href="https://smallstep.com/docs/step-cli/basic-crypto-operations#run-an-offline-x509-certificate-authority">This article</link>
          may also be useful if you want to customize certain aspects of
          certificate generation for your CA.
        '';
```

midchildan commented on nixos/modules/services/security/step-ca.nix:
```diff
@@ -0,0 +1,68 @@
+    services.step-ca = {
+      enable = lib.mkEnableOption "the smallstep certificate authority server";
+      settings = lib.mkOption {
+        type = lib.types.anything;
```
```suggestion
        type = with lib.types; attrsOf anything;
```

midchildan commented on nixos/modules/services/security/step-ca.nix:
```diff
@@ -0,0 +1,68 @@
+        description = "See https://smallstep.com/docs/step-ca/configuration. https://smallstep.com/docs/step-cli/basic-crypto-operations#run-an-offline-x509-certificate-authority may also be useful. The easiest way is probably to use the existing smallstep tools to set up the CA and then import the config using e.g. `builtins.fromJSON`";
+      };
+      intermediatePasswordFile = lib.mkOption {
+        type = lib.types.str;
```
```suggestion
        type = with lib.types; nullOr path;
```

midchildan commented on nixos/modules/services/security/step-ca.nix:
```diff
@@ -0,0 +1,68 @@
+        type = lib.types.str;
+        default = null;
+        example = "/run/keys/smallstep-password";
+        description = "Path to the file containing the password for the intermediate certificate private key. WARNING: make sure to use a quoted absolute path or it is going to be copied to the nix store.";
```
Same here.

```suggestion
        description = ''
          Path to the file containing the password for the intermediate
          certificate private key.

          <warning>
            <para>
              Make sure to use a quoted absolute path instead of a path literal
              to prevent it from being copied to the globally readable Nix
              store.
            </para>
          </warning>
        '';
```

midchildan commented on nixos/modules/services/security/step-ca.nix:
```diff
@@ -0,0 +1,68 @@
+      assertions =
+        [{
+          assertion = !lib.isStorePath cfg.intermediatePasswordFile;
+          message = "`services.step-ca.intermediatePasswordFile` points to a file in the nix store. You should use a quoted absolute path to prevent this.";
```
Ditto.

```suggestion
          message = ''
            <option>services.step-ca.intermediatePasswordFile</option> points to
            a file in the Nix store. You should use a quoted absolute path to
            prevent this.
          '';
```

midchildan commented on nixos/modules/services/security/step-ca.nix:
```diff
@@ -0,0 +1,68 @@
+      systemd.packages = [ pkgs.step-ca ];
+
+      # configuration file indirection is needed to support reloading
+      environment.etc."smallstep-ca.json".source = configFile;
```
I think you want `systemd.services.step-ca.restartTriggers = [ cfg.settings ];`

https://search.nixos.org/options?channel=20.09&show=systemd.services.%3Cname%3E.restartTriggers&from=0&size=50&sort=relevance&query=restartTriggers

aanderse replied:
`Environment = [ "" "SMALLSTEP_HACK_DETECT_CONFIG_CHANGE=${configFile}" ]; # I don't know if there is a cleaner way but this forces the activation script to reload this unit on configuration updates.` - this is what `restartTriggers` is. So you want `systemd.services.step-ca.restartTriggers = [ configFile ];`. What @midchildan has may also work if `cfg.settings` is implicitly cast to a path in the nix store, but off the top of my head I'm not sure.

midchildan replied:
FWIW `restartTriggers` can contain anything that can be converted to a string, though a store path would have the benefit of being significantly shorter than the full contents of `cfg.settings`.

From the commit that introduced `restartTriggers`:

> The triggers are just arbitrary strings that are included in the unit
under X-Restart-Triggers.  The idea is that if they change between reconfigurations, switch-to-configuration will restart the unit because its store path changed.

https://github.com/NixOS/nixpkgs/commit/b606165bd9d4c8a1f82160f67bd541f064ed1ddc

aanderse replied:
Sorry, yes I should have been more specific and clarified that point. Thanks.

midchildan commented on nixos/modules/services/security/step-ca.nix:
```diff
@@ -0,0 +1,68 @@
+
+      systemd.services."step-ca" = {
+        wantedBy = [ "multi-user.target" ];
+        reloadIfChanged = true;
```
Should be unnecessary with `restartTriggers`.

aanderse replied:
It depends if the intention is to restart the service every time configuration changes, as opposed to reloading the service.

midchildan commented:
This module might also benefit from having an `openFirewall` option.
https://search.nixos.org/options?channel=20.09&from=0&size=50&sort=relevance&query=openfirewall

aanderse commented on nixos/modules/services/security/step-ca.nix:
```diff
@@ -0,0 +1,68 @@
+            ""
+            ''${pkgs.step-ca}/bin/step-ca /etc/smallstep-ca.json --password-file ''${CREDENTIALS_DIRECTORY}/intermediate_password''
+          ]; # override upstream
+          ExecReload = [ "" "${pkgs.coreutils}/bin/kill -HUP $MAINPID" ]; # override upstream
```
Might be worth patching the `ExecReload` command in the package :man_shrugging:

mohe2015 replied:
@aanderse https://github.com/NixOS/nixpkgs/pull/112322/files#diff-92af97bdaab3e0156ca22f8ba3be2ad003e7fd5050b1ff8f8fa5fafcfab33f30R31 is this what you meant?

aanderse replied:
@mohe2015 yes.

mohe2015 commented:
@midchildan Thanks a lot for the review and easy-to-apply Github suggestions. I implemented an `openFirewall` option but I don't know if there would be a better way to do this as unfortunately upstream doesn't have a `port` option.

mohe2015 commented:
Can anybody reproduce https://github.com/NixOS/nixpkgs/pull/112322/commits/52ced6297e8a37864c2ee23a683edba035bb9a86 running with `nix build -L /etc/nixos/nixpkgs#step-ca`. I think I'm either too tired or something is really wrong...

Church- commented:
@mohe2015 Unable to reproduce it so far in my local env.

mohe2015 commented:
Ok, I was too tired. using `$src` in the install phase obviously references the non-patched version...
Thanks for testing though @Church-

midchildan commented on nixos/modules/services/security/step-ca.nix:
```diff
@@ -0,0 +1,125 @@
+
+          LoadCredential = "intermediate_password:${cfg.intermediatePasswordFile}"; # DynamicUser means we don't know which user needs access to this file. Therefore let systemd handle this.
+
+          ExecStart = [
+            ""
+            ''${pkgs.step-ca}/bin/step-ca /etc/smallstep-ca.json --password-file ''${CREDENTIALS_DIRECTORY}/intermediate_password''
+          ]; # override upstream
```
Handle the case where `cfg.intermediatePasswordFile` is `null`.
```suggestion
          ExecStart = [
            "" # override upstream
            ("${pkgs.step-ca}/bin/step-ca /etc/ca.json"
              + optionalString (cfg.intermediatePasswordFile != null)
              " --password-file \${CREDENTIALS_DIRECTORY}/intermediate_password")
          ];
```

mohe2015 replied:
In my opinion we shouldn't allow setting this to null as then the private key would be word-readable in the NIx store. The current solution is already not optimal as the config file can contain oauth secrets afaik

midchildan commented on nixos/modules/services/security/step-ca.nix:
```diff
@@ -0,0 +1,125 @@
+          WorkingDirectory = ""; # override upstream
+          ReadWriteDirectories = ""; # override upstream
+
+          LoadCredential = "intermediate_password:${cfg.intermediatePasswordFile}"; # DynamicUser means we don't know which user needs access to this file. Therefore let systemd handle this.
```
Same here.
```suggestion
          # LocalCredentials handles file permission problems arising from the
          # use of DynamicUser.
          LoadCredential =
            lib.OptionalString (cfg.intermediatePasswordFile != null)
            "intermediate_password:${cfg.intermediatePasswordFile}";
```

midchildan reviewed (changes requested):
Just a few things left from me.

midchildan commented on nixos/modules/services/security/step-ca.nix:
```diff
@@ -0,0 +1,125 @@
+            type = lib.types.str;
+            default = "127.0.0.1";
+            example = "0.0.0.0";
+            description = "The address (and port but see <literal>port</literal> option for more details) the certificate authority should listen at.";
```
The defaults should be `null` as well, assuming that it's moved out of `cfg.settings`.

tashian commented on pkgs/tools/security/step-ca/default.nix:
```diff
@@ -2,30 +2,39 @@
   nativeBuildInputs = [ pkg-config ];
 
   buildInputs =
     lib.optional stdenv.isLinux (lib.getDev pcsclite)
```
pcsclite and PCSC are not strictly required to build step-ca from source. They are only required if you want to build it with YubiKey support. The default `make` rule will build a non-CGO version without YubiKey support, but we had a bug in our Makefile around this, which @maraino just fixed in smallstep/certificates#474.

mohe2015 replied:
Thanks but as we run the tests I will keep it for now. I also don't know anything about our go infrastructure so I didn't manage to remove the support anyways. (Removing from `buildInputs` and disabling the tests wasn't enough)

tashian replied:
Do you have an older build? Just to be clear, we're releasing a new version of `step-ca` in the next 24-48 hours that will run `make test` without requiring `pcsclite`. 

So, I think this is really a matter of whether you want the nixOS package to have YubiKey support or not. The default build for `make` is a non-CGO (non-YubiKey supporting) `step-ca`.

mohe2015 replied:
@tashian I found out that CGO is enabled by default on NixOS, I will push a fix soon.

mohe2015 replied:
Done in https://github.com/NixOS/nixpkgs/pull/112322/commits/ec6555c278528705f922a7d517a62b6f105bac86. @midchildan I should probably add a `package` option now?

maraino replied:
@mohe2015 in the next release, if you compile with CGO, besides YubiKey support you will also have PKCS#11 support.

mohe2015 replied:
@maraino Nice! This pull request is currently using the latest `master` commit and it seems like enabling CGO and YubiKey support using the option `securityTokenSupport` works (didn't test with a key but it compiles and runs). Is the name `securityTokenSupport` a sensible one (we could add additional options later)?

maraino replied:
`securityTokenSupport` won't be the right name, as we don't use those devices to 2FA, we use them to store keys, and do sign operations. @tashian what would be a proper name, `hsmSupport` (`hardwareSecurityModuleSupport`)?

maraino replied:
`securityModuleSupport` as the HSM can also be implemented in software (it should only be used for testing purposes)

tashian replied:
I like `hsmSupport` or `pkcs11Support`

midchildan replied:
@mohe2015 

> @midchildan I should probably add a package option now?

I think that would be nice.

As to whether yubikey support should be enabled by default or not, that would depend on the number of users willing to use it. We can disable it for now and enable it later on if it becomes a popular feature, but I personally won't mind if we just enable it now because there's probably already a gazillion packages that depends on pcsclite anyways.

mohe2015 replied:
@tashian @maraino  Any estimations how many users use the yubikey support?

@midchildan We could also enable it by default so users don't have to rebuild. It may be a larger attack surface though although I don't think that's actually relevant.

I personally don't really have an opinion on it but I don't use yubikey support myself.

tashian replied:
@mohe2015 In short, we'll see! We've had a lot of customers who need it for commercial use, mostly because they need FIPS 140 or other security compliance. Not sure how popular PKCS11 will be in open source. We got a lot of traction in the homelab community around our experimental YubiKey PIV support, when we posted our [Tiny CA tutorial](https://smallstep.com/blog/build-a-tiny-ca-with-raspberry-pi-yubikey/) in December.

mohe2015 replied:
I enabled support for now so it's easier to get started as the pcsclite package is build already anyways.

mohe2015 commented on nixos/modules/services/security/step-ca.nix:
```diff
@@ -0,0 +1,125 @@
+        type = lib.types.submodule {
+          freeformType = with lib.types; attrsOf anything;
+
+          options.address = lib.mkOption {
```
`address` would also be a step-ca option. But I think the easiest way is probably to just move both out of `settings` and unconditionally override `settings.address` with `address`:`port`. Was that what you meant? (This would mean both these options are required)

mohe2015 commented:
@midchildan thanks a lot for yet another review! I didn't have much time but I think I incoporated your requested changes. The way I did it may be discussed of course but I tried to keep it as simple as possible.

midchildan commented on nixos/modules/services/security/step-ca.nix:
```diff
@@ -0,0 +1,125 @@
+      systemd.packages = [ cfg.package ];
+
+      # configuration file indirection is needed to support reloading
+      environment.etc."smallstep-ca.json".source = configFile;
```
```suggestion
      environment.etc."smallstep/ca.json".source = configFile;
```

This would allow users to copy certificates from `$STEPPATH/config` to `/etc/smallstep` if they prefer, removing the need to rewrite paths in `ca.json`.

midchildan commented on nixos/modules/services/security/step-ca.nix:
```diff
@@ -0,0 +1,125 @@
+
+          ExecStart = [
+            "" # override upstream
+            "${cfg.package}/bin/step-ca /etc/smallstep-ca.json --password-file \${CREDENTIALS_DIRECTORY}/intermediate_password"
```
Same here.

```suggestion
            "${cfg.package}/bin/step-ca /etc/smallstep/ca.json --password-file \${CREDENTIALS_DIRECTORY}/intermediate_password"
```

mohe2015 commented on pkgs/tools/security/step-ca/default.nix:
```diff
@@ -2,29 +2,43 @@
 buildGoModule rec {
   pname = "step-ca";
-  version = "0.15.6";
+  version = "0.15.7-unstable";
```
We should probably wait for the next release before we merge. This should only take a few days.

mohe2015 commented on nixos/modules/services/security/step-ca.nix:
```diff
@@ -0,0 +1,133 @@
+          ConditionFileNotEmpty = ""; # override upstream
+        };
+        serviceConfig = {
+          Environment = "HOME=$STATE_DIRECTORY";
```
For some reason this is needed for initial startup (I think only without hsm support). I thought systemd would automatically set this but either this doesn't work or something else is wrong.

midchildan replied:
It should be fine, I believe it's standard behavior.

mohe2015 commented:
Checks have issues.

mohe2015 commented:
My usage idea for interested people (Yeah I know I should probably use DNSSEC instead of hijacking my own website's dns for services that are only available in my home):
<details><summary>CLICK ME</summary>
<p>

```nix
{
  security.acme.server = "https://localhost:8443/acme/acme/directory";
  security.acme.email = "root@example.org";
  security.acme.acceptTerms = true;

  environment.etc."nixos/secrets/pi-smallstep-ca".source = ../secrets/pi-smallstep-ca;
  services.step-ca = {
    enable = true;
    address = "127.0.0.1";
    port = 8443;
    intermediatePasswordFile = "/etc/nixos/secrets/pi-smallstep-ca";
    settings = {
      dnsNames = ["localhost"];
      root = ../secrets/root-ca.crt;
      crt = ../secrets/intermediate_ca.crt;
      key = ../secrets/intermediate_ca.key;
      db = {
        type = "badger";
        dataSource = "/var/lib/step-ca/db";
      };
      authority = {
        provisioners = [
          {
            type = "ACME";
            name = "acme";
          }
        ];
      };
    };
  };
  security.pki.certificateFiles = [ ../secrets/root-ca.crt ../secrets/intermediate_ca.crt ];

  services.httpd.adminAddr = "root@example.org";

  # https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/networking/bind.nix
  # networking.resolvconf.useLocalResolver = false; # to unbreak your internet
  services.bind = {
    enable = true;
    zones = [
      {
        name = "rpz";
        master = true;
        file = pkgs.writeText "bind.conf"
        ''
$TTL    604800
@                    IN      SOA     localhost. root.localhost. (
                              5         ; Serial
                         604800         ; Refresh
                          86400         ; Retry
                        2419200         ; Expire
                         604800 )       ; Negative Cache TTL
                             IN      NS localhost.

blog.example.org  A       127.0.0.1
        '';
      }
    ];
    extraOptions = ''
      response-policy { zone "rpz"; };
    '';
  };

  # https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-apps/wordpress.nix
  services.wordpress."blog.example.org" = {
    virtualHost = {
      forceSSL = true;
      enableACME = true;
    };
  };

  services.mysql.package = pkgs.mariadb;
}

```

</p>
</details>


![image](https://user-images.githubusercontent.com/13287984/108259687-b0635b80-7161-11eb-8969-1ecc5cebafe4.png)

Church- commented:
@mohe2015 Ah that's nifty.

Been using my non-declarative module for a while for SSO SSH inside my home and remote infra rather like from their article.

midchildan commented on nixos/modules/services/security/step-ca.nix:
```diff
@@ -0,0 +1,133 @@
+          ConditionFileNotEmpty = ""; # override upstream
+        };
+        serviceConfig = {
+          Environment = "HOME=$STATE_DIRECTORY";
```
The environment variable `STATE_DIRECTORY` may be a colon-separated list of directories, so use this instead to ensure it continues to work in the event an user adds an entry to `systemd.services.step-ca.serviceConfig.StateDirectory`:

```suggestion
          Environment = "HOME=%S/step-ca";
```

`%S` expands to `/var/lib`.

mohe2015 replied:
Good catch!

mohe2015 committed:
step-ca: 0.15.6 -> 0.15.11

mohe2015 committed:
step-ca: Add systemd service file at correct location to get it picked up automatically

mohe2015 committed:
step-ca: Add option to disable HSM support

mohe2015 commented:
@midchildan Thanks for the review. I addressed the requested changes, updated the step-ca version and split up the commits. I hope I split them up correctly but other than that I think this should be ready for a final review now.

midchildan reviewed (approved):
Thanks for your great work! This would need to get merged by people with commit rights, so consider posting to this [Discourse thread](https://discourse.nixos.org/t/prs-already-reviewed/2617) to indicate that this PR has gone through review.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/351

mohe2015 commented:
@midchildan Also thanks a lot for your help in reviewing. You probably invested quite some time as well! Your reviews also were really helpful.

tashian commented:
We're very excited to have such a nice NixOS package for step-ca, thank you @mohe2015 @midchildan!

SuperSandro2000 commented on pkgs/tools/security/step-ca/default.nix:
```diff
@@ -2,29 +2,43 @@
-  buildInputs =
-    lib.optional stdenv.isLinux (lib.getDev pcsclite)
-    ++ lib.optional stdenv.isDarwin PCSC;
+  buildInputs = lib.optional hsmSupport
+    (lib.optional stdenv.isLinux (lib.getDev pcsclite)
+      ++ lib.optional stdenv.isDarwin PCSC);
```
```suggestion
  buildInputs = lib.optional hsmSupport
    (lib.optional stdenv.isLinux pcsclite
      ++ lib.optional stdenv.isDarwin PCSC);
```

mohe2015 replied:
@SuperSandro2000 This leads to
```bash
builder for '/nix/store/ryg7969vam518gicqlq8n0gbyhxyyzsf-step-ca-0.15.8.drv' failed with exit code 1; last 10 log lines:
  github.com/go-piv/piv-go/piv
  # pkg-config --cflags  -- libpcsclite
  Package libpcsclite was not found in the pkg-config search path.
  Perhaps you should add the directory containing `libpcsclite.pc'
  to the PKG_CONFIG_PATH environment variable
  No package 'libpcsclite' found
  pkg-config: exit status 1
  go.step.sm/cli-utils/command/version
  internal/profile
  net/http/pprof
```
What about `pcsclite.dev` instead? I have no idea why it doesn't work though as many other packages do the same.

mohe2015 replied:
I'm currently switching to `optionals` instead of `optional` and it seems to be working somehow...

mohe2015 replied:
Ahh, I got it. This creates a nested list and somehow it worked before.

mohe2015 replied:
I fixed it. Traded some minimal duplication for easier readability. I hope this is fine.

Church- commented:
So when is this planned to get reviewed and merged?

Having been using a permutation of this module for a while I'm excited to see it hit mainline.

@tashian A question since you're here. I happen to use SmallStep at work. Is there a way to manage multiple profiles/CA's easily? Bit of a pain switching between my home and work setup. Don't suppose it's as easy as `step ssh login foo` for work and `step ssh login bar` for home?

Church- commented:
Also a secondary thought @mohe2015 to use this as a ssh CA correctly would we want a secondary step ca service to deal with authenticating to the CA and configuring host ssh certs and the like? Think it'd be an obvious pairing to go with this.

I've rather boringly been managing it somewhat by hand for now.

mohe2015 commented:
@Church- can you elaborate please? I haven't used step-ca for SSH and I don't really understand what you mean exactly.

Church- commented:
@mohe2015 Hey to be clear so like the host based service running in this article: https://smallstep.com/blog/diy-single-sign-on-for-ssh/

It would effectively just be a modification of @tashian's script here: https://gist.github.com/tashian/fde43668cbf6e3227fb13ef51db650b8

mohe2015 commented:
@Church- I read it quite quickly but as far as I understand it only needs one step-ca server with *multiple* provisioners (Google OAuth, AWS, SSHPOP).

tashian commented:
@Church- @mohe2015 

A couple notes on this:
* I developed [a renewal method for X.509 certificates that uses systemd timers](https://smallstep.com/docs/step-ca/certificate-authority-server-production#renewal-using-systemd-timers), and it could be straightforward to adapt this approach to SSH host certificate renewal.
* SSH CA support requires the CA keys to be generated (`--ssh` passed to `step ca init`), and provisioners that can get certs to users and/or hosts. The cloud instance identity provisioners are for host enrollment, the `SSHPOP` provisioner handles host cert renewal (trade the old cert in for a new one), and the OIDC provisioner is for user certificates. The tutorial shows how to set up both user and host CAs and certs for a complete solution, but it's easy to separate these out if you only want one or the other.

tashian commented:
@Church- to answer your question about multiple profiles, sort of yes! You can use `STEPPATH` to set the location of the `step` config folder, which defaults to `$HOME/.step`. The caveat is, our SSH configuration shim doesn't handle multiple profiles well, so if you want to use `step ssh config` to enable for seamless SSO, you will need to change each included SSH config file to add the relevant `STEPPATH`. For example:

```
Match exec "bash -c 'STEPPATH=$HOME/.step-homelab step ssh check-host %h'"
        UserKnownHostsFile "/Users/carl/.step/ssh/known_hosts"
        ProxyCommand bash -c 'STEPPATH=$HOME/.step-homelab step ssh proxycommand --provisioner "okta" %r %h %p'
```

I've opened [an issue about this here](https://github.com/smallstep/certificates/issues/492).

Church- commented:
> @Church- to answer your question about multiple profiles, sort of yes! You can use `STEPPATH` to set the location of the `step` config folder, which defaults to `$HOME/.step`. The caveat is, our SSH configuration shim doesn't handle multiple profiles well, so if you want to use `step ssh config` to enable for seamless SSO, you will need to change each included SSH config file to add the relevant `STEPPATH`. For example:
> 
> ```
> Match exec "bash -c 'STEPPATH=$HOME/.step-homelab step ssh check-host %h'"
>         UserKnownHostsFile "/Users/carl/.step/ssh/known_hosts"
>         ProxyCommand bash -c 'STEPPATH=$HOME/.step-homelab step ssh proxycommand --provisioner "okta" %r %h %p'
> ```
> 
> I've opened [an issue about this here](https://github.com/smallstep/certificates/issues/492).

Awesome thanks!

Church- commented:
> @Church- to answer your question about multiple profiles, sort of yes! You can use `STEPPATH` to set the location of the `step` config folder, which defaults to `$HOME/.step`. The caveat is, our SSH configuration shim doesn't handle multiple profiles well, so if you want to use `step ssh config` to enable for seamless SSO, you will need to change each included SSH config file to add the relevant `STEPPATH`. For example:
> 
> ```
> Match exec "bash -c 'STEPPATH=$HOME/.step-homelab step ssh check-host %h'"
>         UserKnownHostsFile "/Users/carl/.step/ssh/known_hosts"
>         ProxyCommand bash -c 'STEPPATH=$HOME/.step-homelab step ssh proxycommand --provisioner "okta" %r %h %p'
> ```
> 
> I've opened [an issue about this here](https://github.com/smallstep/certificates/issues/492).

Awesome thanks!

> @Church- @mohe2015
> 
> A couple notes on this:
> 
>     * I developed [a renewal method for X.509 certificates that uses systemd timers](https://smallstep.com/docs/step-ca/certificate-authority-server-production#renewal-using-systemd-timers), and it could be straightforward to adapt this approach to SSH host certificate renewal.
> 
>     * SSH CA support requires the CA keys to be generated (`--ssh` passed to `step ca init`), and provisioners that can get certs to users and/or hosts. The cloud instance identity provisioners are for host enrollment, the `SSHPOP` provisioner handles host cert renewal (trade the old cert in for a new one), and the OIDC provisioner is for user certificates. The tutorial shows how to set up both user and host CAs and certs for a complete solution, but it's easy to separate these out if you only want one or the other.

Ah so this is nifty and was around the thinking I was aware of.

@mohe2015 Would we be open to a follow on commit to this module later on adding this as a sub service?
Would be fairly simplistic I think.

mohe2015 commented:
@Church- Do you mean for the renewal-service? If yes - I think this should be done in a separate pull request.

Church- commented:
> @Church- Do you mean for the renewal-service? If yes - I think this should be done in a separate pull request.

Yep exactly, was just wondering if it made sense to make this a sub service to step-ca. I think it does given it's intrinsically linked.

mohe2015 commented:
@Church- I think it's fine if you do it like that first and if somebody has another/better proposal I think this should be possible to change.

This PR should be ready to merge then again.

maraino commented:
@Church- I've posted [here](https://github.com/smallstep/certificates/discussions/493#discussioncomment-395000) the workaround that I use to work with multiple profiles. Basically `alias sssh='ssh -F $(step path)/ssh/config'`

Church- commented:
> @Church- I've posted [here](https://github.com/smallstep/certificates/discussions/493#discussioncomment-395000) the workaround that I use to work with multiple profiles. Basically `alias sssh='ssh -F $(step path)/ssh/config'`

Cool good to know.

Church- commented:
> @Church- I think it's fine if you do it like that first and if somebody has another/better proposal I think this should be possible to change.
> 
> This PR should be ready to merge then again.

Agreed this looks good to merge.

mohe2015 commented:
@ofborg build step-ca

I fixed a fixed-output hash. Should be good to merge when ofborg also builds.

@r-rmcgibbo is it possible to trigger a rebuild?

tashian commented:
Hi @mohe2015, just a thought I had today about what else would make certificate management easier in NixOS. If this belongs on some other forum let me know, I'd love to join in the discussion.

I came across [this blog post from @Xe](https://christine.website/blog/prometheus-grafana-loki-nixos-2020-11-20) about setting up Grafana, Prometheus, loki, and node_exporter using NixOS. They mentioned doing TLS configuration using a custom CA. In their config it looks like they're pulling PEM files from an external HTTPS endpoint during setup.

It got me wondering what it would take to make ACME and shorter-lived certs possible on the service level in NixOS, with automated renewal. Does NixOS have an ACME client that could be composed with services like NGINX that don't have internal ACME support? I think it would be an excellent complement to the `step-ca` package.

For step-ca we have a [TLS cert renewal approach that uses systemd timers](https://smallstep.com/docs/step-ca/certificate-authority-server-production#renewal-using-systemd-timers), but the cert-renewer@ service almost always needs to be overridden to make renewal work for a given service. Certificate files have to be installed into a service-specific location, or the service has to be sent a HUP signal or restarted, etc. It would be amazing if there were some sort of NixOS cert management "sidecar" that just knew how do the right thing for a given service, and could be configured with an ACME CA.

mohe2015 commented:
@tashian I'm short on time right now but there actually already are options for enabling ACME. https://nixos.org/manual/nixos/stable/options.html#opt-services.nginx.virtualHosts._name_.enableACME is the option for nginx (Page loads really slow) and there exists the same option for apache2. As most services use these as reverse proxies almost every service has this builtin and otherwise you can and usually need to put them as a reverse proxy in front.

https://nixos.org/manual/nixos/stable/#module-security-acme-nginx is a bigger example from the official manual and I use step-ca in my test setup the following way:

```nix
{
  security.acme.server = "https://localhost:9443/acme/acme/directory";
  security.acme.email = "root@example.org";
  security.acme.acceptTerms = true;

  services.step-ca = {
    enable = true;
    address = "127.0.0.1";
    port = 9443;
    intermediatePasswordFile = "/etc/nixos/secrets/pi-smallstep-ca"; # FIXME is world readable
    settings = {
      dnsNames = ["localhost"];
      root = ../../../secrets/root_ca.crt;
      crt = ../../../secrets/intermediate_ca.crt;
      key = ../../../secrets/intermediate_ca.key;
      db = {
        type = "badger";
        dataSource = "/var/lib/step-ca/db";
      };
      authority = {
        provisioners = [
          {
            type = "ACME";
            name = "acme";
          }
        ];
      };
    };
  };
}
```
https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/security/acme.nix contains the source of the acme module. The default seems to be checking daily for renewal which can also be configured. I don't know how good support for mTLS is but this would probably need to be implemented into nginx if it isn't already. If there is a simple config flag then you could also easily enable this in your local configuration.

I don't know how well the services work with updating the certificate but that could probably be found out easily.

mohe2015 committed:
Add mohe2015 as maintainer to step-ca and nixos/step-ca

mohe2015 commented:
I added myself as a maintainer and bumped the version to 0.15.10 instead of 0.15.8. There don't seem to be any relevant changes mostly small documentation things.

I would love to get this merged before 21.05 and maybe adding myself as a maintainer *explicitly* helps.

mohe2015 commented on pkgs/tools/security/step-ca/default.nix:
```diff
@@ -2,29 +2,43 @@
-    lib.optional stdenv.isLinux (lib.getDev pcsclite)
-    ++ lib.optional stdenv.isDarwin PCSC;
+    lib.optionals (hsmSupport && stdenv.isLinux) [ pcsclite ] ++
+    lib.optionals (hsmSupport && stdenv.isDarwin) [ PCSC ];
```
Done

mohe2015 commented on maintainers/maintainer-list.nix:
```diff
@@ -6477,6 +6477,16 @@
     githubId = 754512;
     name = "Mogria";
   };
+  mohe2015 = {
```
Done

SuperSandro2000 commented:
@aanderse could you please take a look at the module?

aanderse reviewed (approved):
Great work on the module. Great support from upstream. Team effort. What more could you ask for? 🎆 

There are a ton of comments so I apologize if this has already been discussed... but I just wanted to confirm that everyone understands exactly what these lines implies:

```
        restartTriggers = [ configFile ];
        reloadIfChanged = true; # allow reloads for more service availability
```

To clarify: the service will never automatically restart, even when software is upgraded. The sysadmin must manually restart the service any time the binary is updated if they want the new binary. Based on what the software does I figure I would just double check given the security implications on OS patching.

mohe2015 commented:
> Great work on the module. Great support from upstream. Team effort. What more could you ask for? 🎆 
> 
> There are a ton of comments so I apologize if this has already been discussed... but I just wanted to confirm that everyone understands exactly what these lines implies:
> 
> ```
>         restartTriggers = [ configFile ];
>         reloadIfChanged = true; # allow reloads for more service availability
> ```
> 
> To clarify: the service will never automatically restart, even when software is upgraded. The sysadmin must manually restart the service any time the binary is updated if they want the new binary. Based on what the software does I figure I would just double check given the security implications on OS patching.

@aanderse Oh no, that's bad. Is it possible to get both behaviours so reload on configuration change and restart on binary change? I'm on my phone right now but as far as I remember the documentation of these options is pretty bad currently.

aanderse commented:
Unfortunately it is not...

ping @flokli we should probably do something about this finally.

mohe2015 commented:
@aanderse I removed `reloadIfChanged = true;`. Is this fine now to get merged?

aanderse reviewed (approved):
I approve the module portion of this PR :+1:
